### DISCUSSION TITLE
A regular user has access to all roles and could think he can elevate his rights just after his role has been modified programmatically

### DISCUSSION THREAD

**Discussion by Pachat** [directus/directus#15767] [Archive]
The goal is to allow users to change their roles between some roles (not including Administrator of course)

On the Directus user collection an `active_role` field (Dropdown) has been added with limited choices to some other roles.
In the DB (PostgreSQL), a BEFORE UPDATE trigger then sets `role` = `active_role` on saving.

After saving **and refreshing the page**, everything is OK. The user has the new role.

The problem appears when not refreshing the page, but clicking directly on the user's avatar in the users list. 
The user sees his form and
> is able to click on the `role` (even if no permission to update and show the `role `has been given in any of his roles)
and has access to the list **all the roles including the Administrator role**. 

After refreshing, the `role` becomes again read-only.
May be some caching problem here?

Is there a better way to set a new role programmatically and have the new permissions applied immediately?

--

In fact, if he tries to save the role, because he has no update rights on `role` he will get a `FORBIDDEN` preventing the elevation.
But, it would be nicer 

- to have the `role `field immediately read-only (as update on role is not allowed)
- or, at least, show only the user's `role`.

As `role` is a Directus field, we do not have access

1. to hide it (as it is among the required fields)
2. to filter the roles shown with 
```
{
	"role": {
		"_eq": "$CURRENT_ROLE"
	}
}
```

--

After all, #15761 may solve this.

**Comment by rijkvanzanten** [comment]
> After refreshing, the role becomes again read-only.
> May be some caching problem here?

Yes. Permissions are cached in the system cache, and flushed when the role is changed from Directus. If you're changing the role manually in the database (like with a trigger), make sure to flush the caches 👍🏻

**Reply by Pachat** [reply]
Thanks for the precision. How can caches be flushed after a DB trigger has changed * the role?

\* The role is changed from the value of a custom Directus_users field `active_role`

**Reply by licitdev** [reply]
You may send a POST request or GraphQL mutation to clear caches using an admin account. 
Ref: https://docs.directus.io/reference/system/utilities.html#clear-the-internal-cache

**Reply by Pachat** [reply]
A `regular user` will change his own role following this workflow:

1. The user accesses his profile
2. He clicks on the `Role` Directus_users' field and selects one of the roles he is allowed to use through his Directus_users read filter on his role
3. He clicks on Save
=> It is necessary to flush caches here _while the user changes his own role using the `Role` Directus_users' field and no change are done programmatically through a trigger_?

--

A `manager user` (with no Admin access) will change the role of another `regular user` not by clicking on the `Role` of that user but by selecting a role in a custom Directus_users `active_role` field.
1. The manager user accesses the regular user's profile
2. He clicks on an `active_role` field and selects one of the roles he wants the user to give to the regular user
3. He clicks on Save
4. A DB trigger changes the regular user's role to the value of the active_role
=> It is necessary to flush caches here _while the role is changed programmatically through a trigger but of another user*_?

* That regular user may or not be connected on another terminal.

--

If caches have to be flushed, is it possible to place the POST in a Directus Flow?

**Reply by licitdev** [reply]
The `role` is fetched from the JWT, therefore a page refresh is necessary to get an updated token with the new role.

https://github.com/directus/directus/blob/e4df9b09ba04ff087bfcf44fcd6dc11516b62a81/api/src/middleware/authenticate.ts#L50-L59

No, I believe caches do not have to be flushed. The permissions cache key includes the role, so permissions from a previous role aren't used, provided that the token is refreshed as mentioned above.

https://github.com/directus/directus/blob/e4df9b09ba04ff087bfcf44fcd6dc11516b62a81/api/src/utils/get-permissions.ts#L21-L24

@br41nslug can web sockets be used to monitor changes to the current user's role, triggering a page refresh?

**Reply by Pachat** [reply]
Thanks for the explanation. A page refresh will do.

Just to make things clear to me (for other situations) `POST /utils/cache/clear` acts on the server caches not the client?

**Reply by licitdev** [reply]
> `POST /utils/cache/clear` acts on the server caches not the client

Yes, that is correct. 👍

**Reply by br41nslug** [reply]
> @br41nslug can websockets be used to monitor changes to the current user's role, triggering a page refresh

That is indeed possible 🤔 Although i think invalidating the token with the old user object is probably a more robust solution. This feels related to https://github.com/directus/directus/discussions/15897#discussioncomment-3817309
